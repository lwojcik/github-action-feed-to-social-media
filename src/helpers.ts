import { postToTwitter } from './services/twitter';
import { postToMastodon } from './services/mastodon';
import { FeedItem, PostSubmitStatus, SocialService } from './types';
import { config } from './config';
import { updateMastodonMetadata } from './services/mastodon-metadata';
import { postToDiscord } from './services/discord';
import { postToSlack } from './services/slack';
import { logger } from './logger';

export const stripHTML = (content: string) =>
  content.replace(/(<([^>]+)>)/gi, '');

export const convertArrayToObject = (
  arr: { [key: string]: string }[]
): { [key: string]: string } => {
  const result: { [key: string]: string } = {};

  for (const item of arr) {
    const key = Object.keys(item)[0];
    const value = Object.values(item)[0];
    result[key] = value;
  }

  return result;
};

export const isFeedItemNewer = ({
  feedItem,
  cachedItem,
}: {
  feedItem?: FeedItem;
  cachedItem?: FeedItem;
}) => {
  if (feedItem && cachedItem) {
    return feedItem.published > cachedItem.published;
  }

  return false;
};

export const postToSocialMedia = (params: {
  type: SocialService;
  content: FeedItem;
}) => {
  const { type } = params;
  const { title, link } = params.content;
  const { POST_FORMAT } = config;

  if (!title && !link) {
    logger.notice('Both post title and link are empty. Skipping...');
    return PostSubmitStatus.skipped;
  }

  const postContent = POST_FORMAT.replace('{title}', title || '').replace(
    '{link}',
    link || ''
  );

  switch (type) {
    case SocialService.mastodon:
      return postToMastodon(postContent);
    case SocialService.mastodonMetadata:
      return updateMastodonMetadata(link || '');
    case SocialService.twitter:
      return postToTwitter(postContent);
    case SocialService.discord:
      return postToDiscord(postContent);
    case SocialService.slack:
      return postToSlack(postContent);
    default:
      throw new Error(
        `Unknown social media type: '${type}'. Available types: ${Object.keys(
          SocialService
        )}`
      );
  }
};
