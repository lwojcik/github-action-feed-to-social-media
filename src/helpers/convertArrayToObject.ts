export const convertArrayToObject = (
  arr: { [key: string]: string }[]
): { [key: string]: string } => {
  const result: { [key: string]: string } = {};

  for (const item of arr) {
    const key = Object.keys(item)[0];
    const value = Object.values(item)[0];
    result[key] = value;
  }

  return result;
};
