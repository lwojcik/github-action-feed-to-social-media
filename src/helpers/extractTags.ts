export const extractTags = (postFormat: string) => {
  const regex = /{([A-Za-z0-9:.]+)}/g;
  const matches = postFormat.match(regex);
  if (matches) {
    return matches.map((match) => match.slice(1, -1));
  } else {
    return [];
  }
};
