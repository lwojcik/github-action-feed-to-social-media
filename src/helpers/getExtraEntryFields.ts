import {
  Enclosure,
  ExtraEntryField,
  ExtraEntryProperty,
  FeedItem,
} from '../types';
import { removeUndefinedProps } from './removeUndefinedProps';

export const getExtraEntryFields = (feedEntry: Record<string, any>) => {
  const extraFields: Partial<FeedItem> = {
    [ExtraEntryField.enclosure]: feedEntry[ExtraEntryField.enclosure]
      ? removeUndefinedProps<Enclosure>({
          url:
            feedEntry[ExtraEntryField.enclosure][ExtraEntryProperty.url] ||
            undefined,
          type:
            feedEntry[ExtraEntryField.enclosure][ExtraEntryProperty.type] ||
            undefined,
          length:
            feedEntry[ExtraEntryField.enclosure][ExtraEntryProperty.length] ||
            undefined,
        })
      : undefined,
    [ExtraEntryField.category]: feedEntry[ExtraEntryField.category]?.[
      ExtraEntryProperty.text
    ]
      ? feedEntry[ExtraEntryField.category][ExtraEntryProperty.text]
      : feedEntry[ExtraEntryField.category],
    [ExtraEntryField.pubDate]: feedEntry[ExtraEntryField.pubDate] || undefined,
    [ExtraEntryField.itunesSubtitle]: feedEntry[ExtraEntryField.itunesSubtitle],
    [ExtraEntryField.itunesImage]:
      feedEntry[ExtraEntryField.itunesImage]?.[ExtraEntryProperty.href],
    [ExtraEntryField.itunesExplicit]: feedEntry[ExtraEntryField.itunesExplicit],
    [ExtraEntryField.itunesKeywords]: feedEntry[ExtraEntryField.itunesKeywords],
    [ExtraEntryField.itunesEpisodeType]:
      feedEntry[ExtraEntryField.itunesEpisodeType],
    [ExtraEntryField.itunesDuration]: feedEntry[ExtraEntryField.itunesDuration],
    [ExtraEntryField.itunesEpisode]: feedEntry[ExtraEntryField.itunesEpisode],
  };

  return removeUndefinedProps(extraFields);
};
