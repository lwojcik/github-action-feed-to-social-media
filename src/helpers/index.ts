export * from './convertArrayToObject';
export * from './extractTags';
export * from './formatPostContent';
export * from './isFeedItemNewer';
export * from './logger';
export * from './postToSocialMedia';
export * from './stripHTML';
