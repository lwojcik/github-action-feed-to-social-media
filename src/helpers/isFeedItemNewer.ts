import { FeedItem } from '../types';

export const isFeedItemNewer = ({
  feedItem,
  cachedItem,
}: {
  feedItem?: FeedItem;
  cachedItem?: FeedItem;
}) => {
  if (feedItem && cachedItem) {
    return feedItem.published > cachedItem.published;
  }

  return false;
};
