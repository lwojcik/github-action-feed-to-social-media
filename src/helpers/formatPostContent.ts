import { FeedItem } from '../types';
import { extractTags } from './extractTags';

export const formatPostContent = (content: FeedItem, postFormat: string) => {
  const itemTags = extractTags(postFormat);

  return itemTags.reduce((acc, itemTag) => {
    const nestedTag = itemTag.split('.');

    if (nestedTag.length === 2) {
      return acc.replace(
        `{${itemTag}}`,
        (content[nestedTag[0]] as Record<string, any>)[nestedTag[1]] as string
      );
    }

    return acc.replace(`{${itemTag}}`, content[itemTag] as string);
  }, postFormat);
};
