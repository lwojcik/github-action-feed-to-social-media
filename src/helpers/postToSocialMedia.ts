import { FeedItem, SocialService } from '../types';
import { updateMastodonMetadata } from '../services/mastodon-metadata';
import { postToDiscord } from '../services/discord';
import { postToSlack } from '../services/slack';
import { postToTwitter } from '../services/twitter';
import { postToMastodon } from '../services/mastodon';
import { config } from '../config';
import { formatPostContent } from './formatPostContent';
import { logger } from '../logger';

export const postToSocialMedia = (params: {
  type: SocialService;
  content: FeedItem;
}) => {
  const { content, type } = params;
  const { link } = params.content;
  const { POST_FORMAT } = config;

  if (!link) {
    logger.notice(
      'Post link is empty. If you enabled Mastodon metadata update, it will not be triggered.'
    );
  }

  const formattedPost = formatPostContent(content, POST_FORMAT);

  switch (type) {
    case SocialService.mastodon:
      return postToMastodon(formattedPost);
    case SocialService.mastodonMetadata:
      return updateMastodonMetadata(link || '');
    case SocialService.twitter:
      return postToTwitter(formattedPost);
    case SocialService.discord:
      return postToDiscord(formattedPost);
    case SocialService.slack:
      return postToSlack(formattedPost);
    default:
      throw new Error(
        `Unknown social media type: '${type}'. Available types: ${Object.keys(
          SocialService
        )}`
      );
  }
};
