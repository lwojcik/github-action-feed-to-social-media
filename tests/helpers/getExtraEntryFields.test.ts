import { getExtraEntryFields } from '../../src/helpers/getExtraEntryFields';

describe('getExtraEntryFields', () => {
  test('returns expected extra fields when feedEntry has all required properties', () => {
    const feedEntry = {
      enclosure: {
        '@_url': 'https://example.com/audio.mp3',
        '@_type': 'audio/mpeg',
        '@_length': 1024,
      },
      category: {
        '@_text': 'Technology',
      },
      pubDate: '2023-07-19',
      'itunes:subtitle': 'Episode subtitle',
      'itunes:image': 'https://example.com/image.jpg',
      'itunes:explicit': false,
      'itunes:keywords': 'keyword1,keyword2',
      'itunes:episodeType': 'full',
      'itunes:duration': '00:30:00',
      'itunes:episode': 1,
    };

    const result = getExtraEntryFields(feedEntry);
    expect(result).toMatchSnapshot();
  });

  test('returns expected extra fields when feedEntry has missing properties', () => {
    const feedEntry = {
      enclosure: {
        url: 'https://example.com/audio.mp3',
      },
      category: {},
    };

    const result = getExtraEntryFields(feedEntry);
    expect(result).toMatchSnapshot();
  });

  test('returns expected extra fields when feedEntry has undefined properties', () => {
    const feedEntry = {
      enclosure: undefined,
      category: undefined,
    };

    const result = getExtraEntryFields(feedEntry);
    expect(result).toMatchSnapshot();
  });

  test('returns expected extra fields when feedEntry has incomplete properties', () => {
    const feedEntry = {};

    const result = getExtraEntryFields(feedEntry);
    expect(result).toMatchSnapshot();
  });
});
