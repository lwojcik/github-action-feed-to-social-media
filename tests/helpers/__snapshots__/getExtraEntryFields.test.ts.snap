// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getExtraEntryFields returns expected extra fields when feedEntry has all required properties 1`] = `
{
  "category": "Technology",
  "enclosure": {
    "length": 1024,
    "type": "audio/mpeg",
    "url": "https://example.com/audio.mp3",
  },
  "itunes:duration": "00:30:00",
  "itunes:episode": 1,
  "itunes:episodeType": "full",
  "itunes:explicit": false,
  "itunes:keywords": "keyword1,keyword2",
  "itunes:subtitle": "Episode subtitle",
  "pubDate": "2023-07-19",
}
`;

exports[`getExtraEntryFields returns expected extra fields when feedEntry has incomplete properties 1`] = `{}`;

exports[`getExtraEntryFields returns expected extra fields when feedEntry has missing properties 1`] = `
{
  "category": {},
  "enclosure": {},
}
`;

exports[`getExtraEntryFields returns expected extra fields when feedEntry has undefined properties 1`] = `{}`;
